api/dps/server_integration_test.go

```diff
@@ -0,0 +1,1024 @@
+		assert.Equal(t, txID[:], resp.TransactionID)
+
+		wantData, err := codec.Marshal(results[0])
+		require.NoError(t, err)
+		assert.Equal(t, wantData, resp.Data)
+	})
+
+	t.Run("handles indexer failure on Result", func(t *testing.T) {
+		t.Parallel()
+
+		codec, err := zbor.NewCodec()
+		require.NoError(t, err)
+
+		db := helpers.InMemoryDB(t)
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+
+		// No data is written in the database, so the index should fail to retrieve anything.
+
+		server := dps.NewServer(reader, codec)
+
+		req := &dps.GetResultRequest{
+			TransactionID: txID[:],
+		}
+		_, err = server.GetResult(context.Background(), req)
+
+		assert.Error(t, err)
+	})
+}
+
+func TestIntegrationServer_GetSeal(t *testing.T) {
+	seals := mocks.GenericSeals(4)
+	sealID := seals[0].ID()
+
+	t.Run("nominal case", func(t *testing.T) {
+		t.Parallel()
+
+		codec, err := zbor.NewCodec()
+		require.NoError(t, err)
+
+		db := helpers.InMemoryDB(t)
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+		writer := index.NewWriter(db, storage)
+
+		// Insert mock data in database.
+		require.NoError(t, writer.Seals(mocks.GenericHeight, seals))
+		require.NoError(t, writer.Close())
+
+		server := dps.NewServer(reader, codec)
+
+		req := &dps.GetSealRequest{
+			SealID: sealID[:],
+		}
+		resp, err := server.GetSeal(context.Background(), req)
+
+		require.NoError(t, err)
+		assert.Equal(t, sealID[:], resp.SealID)
+
+		wantData, err := codec.Marshal(seals[0])
+		require.NoError(t, err)
+		assert.Equal(t, wantData, resp.Data)
+	})
+
+	t.Run("handles indexer failure on Seal", func(t *testing.T) {
+		t.Parallel()
+
+		codec, err := zbor.NewCodec()
+		require.NoError(t, err)
+
+		db := helpers.InMemoryDB(t)
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+
+		// No data is written in the database, so the index should fail to retrieve anything.
+
+		server := dps.NewServer(reader, codec)
+
+		req := &dps.GetSealRequest{
+			SealID: sealID[:],
+		}
+		_, err = server.GetSeal(context.Background(), req)
+
+		assert.Error(t, err)
+	})
+}
+
+func TestIntegrationServer_ListSealsForHeight(t *testing.T) {
+	seals := mocks.GenericSeals(4)
+	height := mocks.GenericHeight
+
+	t.Run("nominal case", func(t *testing.T) {
+		t.Parallel()
+
+		codec, err := zbor.NewCodec()
+		require.NoError(t, err)
+
+		db := helpers.InMemoryDB(t)
+
+		storage := storage.New(codec)
+		reader := index.NewReader(db, storage)
+		writer := index.NewWriter(db, storage)
+
+		// Insert mock data in database.
+		require.NoError(t, writer.Seals(height, seals))
+		require.NoError(t, writer.Close())
+
+		server := dps.NewServer(reader, codec)
+
+		req := &dps.ListSealsForHeightRequest{
+			Height: mocks.GenericHeight,
+		}
+		resp, err := server.ListSealsForHeight(context.Background(), req)
+
+		require.NoError(t, err)
+		assert.Equal(t, mocks.GenericHeight, resp.Height)
+
+		for _, collection := range seals {
```

Comment:
Same as before, loop variable wasn't updated, and we need to check the total length of the ID slice.